### PR TITLE
Combine fewsnet & Global IPC processing

### PR DESCRIPTION
**Pull request by Tinkaa**
- Combine the previous two files for processing FewsNet into one (sadly makes it hard to compare changes due to being new file)
- Create Global IPC processing .py script

Think both scripts still need some refactoring but first step

### CODE REVIEW

**Comment by Tinkaa** [comment]
Also added a .py to combine the fewsnet and global ipc data and compute if triggers are met

**Comment by joseepoirier** [comment]
Can you walk me through your logic for combining the two? Ie would it be simpler to have two .py files, one per source?

File: IPC_computetrigger.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    df["year"] = df["date"].dt.year
+    df["month"] = df["date"].dt.month
+
+    df["Source"] = source
+
+    df = df.rename(columns={"adjusted_population": "pop_Country", "ADM1_EN": "ADMIN1"})
+
+    # calculate percentage of population per status and level
+    for status in ["CS", "ML1", "ML2"]:
+        for i in range(1, 6):
```
Comment (joseepoirier) [inline IPC_computetrigger.py:27]: why range(1,6)? (please include answer in code as comment: eg: # number of IPC index levels)

File: IPC_computetrigger.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+def add_columns(df, source):
+    df["date"] = pd.to_datetime(df["date"])
+    df["year"] = df["date"].dt.year
+    df["month"] = df["date"].dt.month
+
+    df["Source"] = source
+
+    df = df.rename(columns={"adjusted_population": "pop_Country", "ADM1_EN": "ADMIN1"})
+
+    # calculate percentage of population per status and level
```
Comment (joseepoirier) [inline IPC_computetrigger.py]: "status" might be better replaced with something like "timepoint"

File: IPC_computetrigger.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    for status in ["CS", "ML1", "ML2"]:
```
Comment (joseepoirier) [inline IPC_computetrigger.py]: Are these headers used by both FewsNet and Global?

File: IPC_computetrigger.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    ):
+        return 1
+    else:
+        return 0
+
+
+def get_trigger_increase(row, status, level, perc):
+    """
+    Return 1 if for row percentage in >="level" at status minus percentage in >="level" currently (CS) is expected to be larger than perc
+    For Global IPC the population analysed in ML2 is sometimes different than in CS. That is why we work dirrectly with percentages and not anymore with (pop status level+ - pop CS level+) / pop CS
```
Comment (joseepoirier) [inline IPC_computetrigger.py]: I don't understand this comment.

File: IPC_computetrigger.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    """
+    # range till 6 cause 5 is max level
+    cols_perc_ml = [f"perc_{status}_{l}" for l in range(level, 6)]
+    cols_perc_cs = [f"perc_CS_{l}" for l in range(level, 6)]
+    if row[["pop_CS", f"pop_{status}"]].isnull().values.any():
+        return np.nan
+    if row[cols_perc_ml].sum() == 0:
```
Comment (joseepoirier) [inline IPC_computetrigger.py]: should this specify which ML?

File: IPC_computetrigger.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        return 0
+    if round(row[cols_perc_ml].sum() - row[cols_perc_cs].sum()) >= perc:
```
Comment (joseepoirier) [inline IPC_computetrigger.py]: Same should this say ML1 or ML2?

File: IPC_computetrigger.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    parameters = parse_yaml(config_file)[country_iso3]
+    country = parameters["country_name"]
+    admin_level = parameters["admin_level"]
+    processed_fews_path = parameters["processed_fews_path"]
+    processed_globalipc_path = parameters["processed_globalipc_path"]
+
+    ipc_cols = [
+        f"{status}_{i}"
+        for status in ["CS", "ML1", "ML2"]
+        for i in [1, 2, 3, 4, 5, "3p", "2m"]
```
Comment (joseepoirier) [inline IPC_computetrigger.py:136]: not clear what "3p" and "2m" refer to?

File: IPC_computetrigger.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    ]
+    pop_cols = [
+        f"pop_{status}" for status in ["CS", "ML1", "ML2", f"ADMIN{admin_level}"]
+    ]
+
+    # initialize dataframes such that can later check if they are filled with data
+    df_fewss = None
+    df_gipcs = None
+
+    if os.path.exists(processed_fews_path):
```
Comment (joseepoirier) [inline IPC_computetrigger.py:153]: should we include a note on the expected format of the csv?

File: IPC_computetrigger.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        df_comb = df_gipcs
+        df_comb_trig = compute_trigger(df_comb)
+        logger.warning("No FewsNet data found")
+
+    else:
+        df_comb_trig = pd.DataFrame()
+        logger.warning("No data found")
+
+    RESULT_FOLDER = f"{country}/Data/IPC_trigger/"
+    Path(RESULT_FOLDER).mkdir(parents=True, exist_ok=True)
```
Comment (joseepoirier) [inline IPC_computetrigger.py:190]: should mkdir be in this .py if it is intended to be run every time a forecast update is available?

File: IPC_computetrigger.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    df_comb_trig.to_csv(f"{RESULT_FOLDER}trigger_results.csv", index=False)
```
Comment (joseepoirier) [inline IPC_computetrigger.py]: Will this overwrite or append results? if append, we should add a timestamp of when the script was run.

File: process_ipc.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
+import logging
```
Comment (joseepoirier) [inline process_ipc.py]: filename should be more specific: process_ipcglobal.py

File: process_ipc.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+        country_iso3: string with iso3 code
+        config_file: path to config file
+    """
+    parameters = parse_yaml(config_file)[country_iso3]
+    country = parameters["country_name"]
+    admin2_shp = parameters["path_admin2_shp"]
+    globalipc_file = parameters["ipc_path"]
+
+    SHP_PATH = f"{country}/Data/{admin2_shp}"
+    IPC_PATH = f"{country}/Data/{globalipc_file}"
```
Comment (joseepoirier) [inline process_ipc.py]: GLOBALIPC_PATH (or IPCGLOBAL_PATH)?

File: IPC_computetrigger.py (hunk 2 above)
Comment (Tinkaa) [reply IPC_computetrigger.py]: Changed to "period"

File: IPC_computetrigger.py (hunk 3 above)
Comment (Tinkaa) [reply IPC_computetrigger.py]: No, only FewsNet but it the original column names are not found, nothing happens. But can be done nicer, added a TODO

File: IPC_computetrigger.py (hunk 5 above)
Comment (Tinkaa) [reply IPC_computetrigger.py]: The ML is specified by "status". Is that what you mean?

File: IPC_computetrigger.py (hunk 6 above)
Comment (Tinkaa) [reply IPC_computetrigger.py]: Same here, the cols_perc_ml contains the ML1 or ML2 columns, depending on the value of "status" that is inputted to the function. Changed "cols_perc_ml" to "cols_perc_proj"

File: IPC_computetrigger.py (hunk 9 above)
Comment (Tinkaa) [reply IPC_computetrigger.py:190]: the "exist_ok=True" argument causes the dir only to be made when it doesn't exist yet so would say that is fine. What would you suggest?

File: IPC_computetrigger.py (hunk 10 above)
Comment (Tinkaa) [reply IPC_computetrigger.py]: It will overwrite. Assuming here a workflow where dates will be added to the input files as time progresses

**Comment by Tinkaa** [comment]
> Can you walk me through your logic for combining the two? Ie would it be simpler to have two .py files, one per source?

Of course! Since the calculation of whether the trigger is met is the same for both sources, my logic was that it is easier to keep that in one file and load the two sources. Also, to have then one final output file with all trigger values seemed easier to me for analysis and e.g. dashboard development

But really not a big preference, could also move the trigger functions to utils and import them to the two .py files. 

@joseepoirier Do you have a preference?

**Comment by Tinkaa** [comment]
@joseepoirier assuming this is okay, will merge

**Pull request merged**